lib/galaxy/webapps/galaxy/api/workflows.py

```diff
@@ -551,10 +556,129 @@ def build_module(self, trans, payload={}):
             'config_form'       : module.get_config_form(),
             'post_job_actions'  : module.get_post_job_actions(inputs)
         }
+        
+    @expose_api
+    def get_tool_predictions(self, trans, topk=10, max_seq_len=25, payload={}):
+        """
+        POST /api/workflows/get_tool_predictions
+        Fetch predicted tools for a workflow
+        """
+        trained_model_path = trans.app.config.model_path
+        tool_sequence = payload.get('tool_sequence', "")
+        if 'tool_sequence' not in payload or trained_model_path is None:
+            return
+        else:
+            r_file = requests.get(trained_model_path)
+            model_save_path = os.path.join(os.getcwd(), "database/trained_model.hdf5")
+            if not os.path.exists(model_save_path):
+                with open(model_save_path, 'wb') as fl:
+                    fl.write(r_file.content)
+            m_file = h5py.File(model_save_path, 'r')
+            toolbx = trans.app.toolbox.to_dict(trans, in_panel=True)
+            all_tools = dict()
+            for category in toolbx:
+                if 'elems' in category:
+                    for item in category['elems']:
+                        if "name" in item and "id" in item:
+                            all_tools[item["id"]] = item["name"]
+            K.clear_session()
+            # retrieve all datasets for creating model and prediction
+            model_config = json.loads(m_file.get('model_config').value)
+            dictionary = json.loads(m_file.get('data_dictionary').value)
+            reverse_dictionary = dict((v,k) for k,v in dictionary.items())
+            compatibile_tools = json.loads(m_file.get('compatible_tools').value)
+            loaded_model = model_from_json(model_config)
```

Comment:
How long does this loading process take? My gut worries about about all of these json loads on every request. But I guess you are doing it this way to make it more self contained?

Reply:
Sometimes the predicted tools are not compatible (they do not connect to the previous tool in the workflow). We should remove those 'incorrect/incompatible' tools from the display.

Reply:
This doesn't really answer my question?

Reply:
What exactly is your question?

Reply:
Hi ... I computed the time taken for all `json.loads` in this tool prediction API
```
import time
s_time = time.time()
model_config = json.loads(trained_model.get('model_config').value)
dictionary = json.loads(trained_model.get('data_dictionary').value)
compatibile_tools = json.loads(trained_model.get('compatible_tools').value)
e_time = time.time()
print("Time elapsed for json parsing: %.4f seconds" % (e_time - s_time))
```

I computed for 5 predictions:
```
Time elapsed for json parsing: 0.0067 seconds
Time elapsed for json parsing: 0.0046 seconds
Time elapsed for json parsing: 0.0052 seconds
Time elapsed for json parsing: 0.0094 seconds
Time elapsed for json parsing: 0.0044 seconds
```

The loading time is in the range of `5 ms`. But, yes it would be good to load this model once and for successive prediction tasks, the model remains available. I can work on this.